lib/spack/llnl/util/argparsewriter.py

```diff
@@ -10,195 +10,345 @@
+        return ''
 
     def begin_optionals(self):
-        self.line()
-        self.line('**Optional arguments**\n')
+        return '\n**Optional arguments**\n\n'
 
     def optional(self, opts, help):
-        self.line('``%s``' % opts)
-        self.line('  %s\n' % help)
+        return """\
+``{0}``
+  {1}
+
+""".format(opts, help)
+
+    def end_optionals(self):
+        return ''
 
     def begin_subcommands(self, subcommands):
-        self.line()
-        self.line('**Subcommands**\n')
-        self.line('.. hlist::')
-        self.line('   :columns: 4\n')
+        string = """
+**Subcommands**
+
+.. hlist::
+   :columns: 4
+
+"""
 
         for cmd in subcommands:
             prog = cmd.prog
             if self.strip_root_prog:
                 prog = re.sub(r'^[^ ]* ', '', prog)
 
-            self.line('   * :ref:`%s <%s>`'
-                      % (prog, cmd.prog.replace(' ', '-')))
-        self.line()
+            string += '   * :ref:`{0} <{1}>`\n'.format(
+                prog, cmd.prog.replace(' ', '-'))
+
+        return string + '\n'
+
+
+class ArgparseCompletionWriter(ArgparseWriter):
+    """Write argparse output as shell programmable tab completion functions."""
+
+    def format(self, prog, description, usage,
+               positionals, optionals, subcommands):
+        """Returns the string representation of a single node in the
+        parser tree.
+
+        Override this in subclasses to define how each subcommand
+        should be displayed.
+
+        Parameters:
+            prog (str): the command name
+            description (str): the command description
+            usage (str): the command usage
+            positionals (list): list of positional arguments
+            optionals (list): list of optional arguments
+            subcommands (list): list of subcommand parsers
+
+        Returns:
+            str: the string representation of this subcommand
+        """
+
+        assert optionals  # we should always at least have -h, --help
+        assert not (positionals and subcommands)  # one or the other, not both
+
+        # We only care about the arguments/flags, not the help messages
+        if positionals:
+            positionals, _ = zip(*positionals)
+        optionals, _, _ = zip(*optionals)
+        subcommands = [
+            subcommand.prog.split(' ')[-1] for subcommand in subcommands
+        ]
+
+        # Flatten lists of lists
+        optionals = [x for xx in optionals for x in xx]
+
+        return self.start_function(prog) + \
+            self.body(positionals, optionals, subcommands) + \
+            self.end_function(prog)
+
+    def start_function(self, prog):
+        """Returns the syntax needed to begin a function definition.
+
+        Parameters:
+            prog (str): the command name
+
+        Returns:
+            str: the function definition beginning
+        """
+        name = prog.replace('-', '_').replace(' ', '_')
+        return '_{0} () {{'.format(name)
```

Comment:
I went with the POSIX-compliant function definition so this won't need to be changed for most shells.